### PR TITLE
49722 initiate creation large media exchange set part1

### PR DESCRIPTION
**Pull request by shailen14599**
Code Changes for Initial large media folder and media.txt part 1
Unit Tests Changes
Functional Tests added

### CODE REVIEW

**Review by sumod13943 — APPROVED** [review]
LGTM

**Review by ajasonca — CHANGES_REQUESTED** [review]
A few comments for you to look at

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common/Helpers/AzureBlobStorageService.cs [hunk 1]
```diff
@@ -133,10 +133,14 @@ public Task<SalesCatalogueProductResponse> DownloadSalesCatalogueResponse(string
...
                 return (mediumExchangeSetInstance.GetInstanceNumber(storageConfig.Value.MediumExchangeSetInstance), ExchangeSetType.mxs);
             }
-            else
+            else if (fileSizeInMB > storageConfig.Value.LargeExchangeSetSizeInMB && fileSizeInMB <= storageConfig.Value.LargeMediaExchangeSetSizeInMB)
             {
                 return (largeExchangeSetInstance.GetInstanceNumber(storageConfig.Value.LargeExchangeSetInstance), ExchangeSetType.lxs);
             }
+            else
+            {
+                return (largeExchangeSetInstance.GetInstanceNumber(1), ExchangeSetType.lxs);
```
[the remark this answers is not in the document]
Comment (harshal11869) [reply UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common/Helpers/AzureBlobStorageService.cs:142]: For POS we are using same Large Exchange Set Instance-1 we need to use it as it is, and large media exchange set comes under lxs only. Just the size of exchange set and folder structure will be different hence not changed it.

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.FulfilmentService/FulfilmentServiceJob.cs [hunk 2]
```diff
@@ -49,13 +53,17 @@ public async Task ProcessQueueMessage([QueueTrigger("%ESSFulfilmentStorageConfig
             string homeDirectoryPath = configuration["HOME"];
             string currentUtcDate = DateTime.UtcNow.ToString("ddMMMyyyy");
             string batchFolderPath = Path.Combine(homeDirectoryPath, currentUtcDate, fulfilmentServiceQueueMessage.BatchId);
+            double fileSizeInMb = CommonHelper.ConvertBytesToMegabytes(fulfilmentServiceQueueMessage.FileSize);
+
+            CommonHelper.IsPeriodicOutputService = fileSizeInMb > periodicOutputServiceConfiguration.Value.LargeMediaExchangeSetSizeInMB;
```
[the remark this answers is not in the document]
Comment (harshal11869) [reply UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.FulfilmentService/FulfilmentServiceJob.cs:58]: Actually we are going to use this IsPeriodicOutputService flag for further processing while creating large media exchange set e.g. Downloading ENCs Files

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.FulfilmentService/Services/FulfilmentDataService.cs [hunk 3]
```diff
@@ -47,12 +52,29 @@ public class FulfilmentDataService : IFulfilmentDataService
...
+            this.fileSystemHelper = fileSystemHelper;
+            this.periodicOutputServiceConfiguration = periodicOutputServiceConfiguration;
         }
 
         public async Task<string> CreateExchangeSet(SalesCatalogueServiceResponseQueueMessage message, string currentUtcDate)
         {
             DateTime createExchangeSetTaskStartedAt = DateTime.UtcNow;
             string homeDirectoryPath = configuration["HOME"];
+
+            if (CommonHelper.IsPeriodicOutputService)
```
[the remark this answers is not in the document]
Comment (harshal11869) [reply UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.FulfilmentService/Services/FulfilmentDataService.cs]: As suggested created new method (CreateLargeExchangeSet) to generate Large Media Exchange Set.

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.FulfilmentService/Services/FulfilmentAncillaryFiles.cs [hunk 4]
```diff
@@ -251,5 +253,32 @@ private string GetIssueAndUpdateDate(Dates dateResponse)
             string ISDT = dateResponse.IssueDate.ToString("yyyyMMdd");
             return (UADT != null) ? $"UADT={UADT},ISDT={ISDT};" : $"ISDT={ISDT};";
         }
+
+        public async Task<bool> CreateMediaFile(string batchId, string folderpath, string correlationId, string exchangeset)
```
[the remark this answers is not in the document]
Comment (harshal11869) [reply UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.FulfilmentService/Services/FulfilmentAncillaryFiles.cs]: Resolved.

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.FulfilmentService/Services/FulfilmentAncillaryFiles.cs [hunk 5]
```diff
@@ -23,12 +23,14 @@ public class FulfilmentAncillaryFiles : IFulfilmentAncillaryFiles
         private readonly IOptions<FileShareServiceConfiguration> fileShareServiceConfig;
         private readonly IFileSystemHelper fileSystemHelper;
         private readonly int crcLength = 8;
+        private readonly IOptions<PeriodicOutputServiceConfiguration> periodicOutputServiceConfiguration;
```
[the remark this answers is not in the document]
Comment (harshal11869) [reply UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.FulfilmentService/Services/FulfilmentAncillaryFiles.cs]: Resolved.

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.FulfilmentService/Services/FulfilmentAncillaryFiles.cs [hunk 6]
```diff
@@ -251,5 +253,32 @@ private string GetIssueAndUpdateDate(Dates dateResponse)
...
+        public async Task<bool> CreateMediaFile(string batchId, string folderpath, string correlationId, string exchangeset)
+        {
+            bool checkMediaFileCreated = false;
+            if (!string.IsNullOrWhiteSpace(folderpath))
+            {
+                string mediaFilePath = Path.Combine(folderpath, periodicOutputServiceConfiguration.Value.LargeExchangeSetMediaFileName);
+                fileSystemHelper.CheckAndCreateFolder(folderpath);
+                int substringvalue = 2;
+                int weekNumber = CommonHelper.GetCurrentWeekNumber(DateTime.UtcNow);
+                var mediaFileContent = String.Format("GBWK{0:D2}_{1}   {2:D4}{3:D2}{4:D2}BASE      M0{5}X02",
```
[the remark this answers is not in the document]
Comment (harshal11869) [reply UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.FulfilmentService/Services/FulfilmentAncillaryFiles.cs]: As suggested used string interpolation for constructing string for MEDIA.TXT content.

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.FulfilmentService/Services/FulfilmentAncillaryFiles.cs [hunk 7]
```diff
@@ -251,5 +253,32 @@ private string GetIssueAndUpdateDate(Dates dateResponse)
...
         }
+
+        public async Task<bool> CreateMediaFile(string batchId, string folderpath, string correlationId, string exchangeset)
+        {
+            bool checkMediaFileCreated = false;
+            if (!string.IsNullOrWhiteSpace(folderpath))
+            {
+                string mediaFilePath = Path.Combine(folderpath, periodicOutputServiceConfiguration.Value.LargeExchangeSetMediaFileName);
+                fileSystemHelper.CheckAndCreateFolder(folderpath);
+                int substringvalue = 2;
```
[the remark this answers is not in the document]
Comment (harshal11869) [reply UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.FulfilmentService/Services/FulfilmentAncillaryFiles.cs]: Resolved

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.FulfilmentService/Services/FulfilmentDataService.cs [hunk 8]
```diff
@@ -47,12 +52,29 @@ public class FulfilmentDataService : IFulfilmentDataService
             this.fulfilmentSalesCatalogueService = fulfilmentSalesCatalogueService;
             this.fulfilmentCallBackService = fulfilmentCallBackService;
             this.monitorHelper = monitorHelper;
+            this.fileSystemHelper = fileSystemHelper;
+            this.periodicOutputServiceConfiguration = periodicOutputServiceConfiguration;
```
[the remark this answers is not in the document]
Comment (harshal11869) [reply UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.FulfilmentService/Services/FulfilmentDataService.cs]: Resolved.

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.FulfilmentService/Services/FulfilmentDataService.cs [hunk 9]
```diff
@@ -254,5 +283,15 @@ public async Task<SalesCatalogueDataResponse> GetSalesCatalogueDataResponse(stri
             SalesCatalogueDataResponse salesCatalogueTypeResponse = await fulfilmentSalesCatalogueService.GetSalesCatalogueDataResponse(batchId, correlationId);
             return salesCatalogueTypeResponse;
         }
+
+        public async Task CreatePosFolderStructure(string largeMediaExchangeSetPath)
```
[the remark this answers is not in the document]
Comment (harshal11869) [reply UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.FulfilmentService/Services/FulfilmentDataService.cs:287]: Resolved.

**Review by ajasonca — APPROVED** [review]

**Pull request merged**